File: 2023_algorithms/golang/quick_sort.go
package goimpls

import (
// "fmt"
)

func partition(nums []int, low, high int) int {
	if low >= high {
		return low
	}

	pivot, value := low, nums[high]

	for i := low; i < high; i++ {
		if nums[i] <= value {
			// fmt.Printf("~~~ i=%d, index=%d, nums[i]=%d\n", i, index, nums[i])
			if pivot != i {
				nums[pivot], nums[i] = nums[i], nums[pivot]
			}
			pivot += 1
		}
	}

	nums[pivot], nums[high] = nums[high], nums[pivot]

	return pivot
}

func qs(nums []int, low, high int) {
	if low >= high {
		return
	}

	pivot := partition(nums, low, high)
	// fmt.Println("~~>", nums, low, high, pivot)

	if low+1 < pivot {
		qs(nums, low, pivot-1)
	}

	if pivot+1 < high {
		qs(nums, pivot+1, high)
	}
}

func QuckSort(nums []int) {
	if len(nums) < 2 {
		return
	}

	qs(nums, 0, len(nums)-1)
}

File: 2023_algorithms/rust/basics/src/sorting/quick_sort.rs
pub fn quick_sort<T: Copy + Ord>(arr: &mut [T]) {
    fn partition<T: Copy + Ord>(arr: &mut [T], low: usize, high: usize) -> usize {
        let mut pivot = low;
        let value = arr[high];

        for i in low..high {
            if arr[i] <= value {
                if pivot != i {
                    arr.swap(pivot, i);
                }
                pivot += 1;
            }
        }

        arr.swap(pivot, high);
        pivot
    }

    fn qs<T: Copy + Ord>(arr: &mut [T], low: usize, high: usize) {
        if low >= high {
            return;
        }

        let pivot = partition(arr, low, high);

        if low + 1 < pivot {
            qs(arr, low, pivot - 1);
        }

        if pivot + 1 < high {
            qs(arr, pivot + 1, high);
        }
    }

    if arr.len() < 2 {
        return;
    }

    qs(arr, 0, arr.len() - 1);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn t_quick_sort() {
        let mut arr = vec![3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];
        quick_sort(&mut arr);
        assert_eq!(arr, vec![1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9])
    }
}
